File: lib/kotlin/gateway/bin/main/dev/bridgething/gateway/UUIDBytes.kt
package dev.bridgething.gateway

import java.nio.ByteBuffer
import java.util.UUID

/**
 * UUIDs travel on the wire as 16-byte msgpack `bin`. The schema-generated types
 * expose those fields as `ByteArray`; these helpers translate to and from
 * `java.util.UUID` at field boundaries.
 */

public fun uuidFromBytes(bytes: ByteArray): UUID {
  require(bytes.size == 16) { "expected 16 bytes for a UUID, got ${bytes.size}" }
  val buf = ByteBuffer.wrap(bytes)
  val msb = buf.long
  val lsb = buf.long
  return UUID(msb, lsb)
}

public fun UUID.toBytes(): ByteArray {
  val buf = ByteBuffer.allocate(16)
  buf.putLong(mostSignificantBits)
  buf.putLong(leastSignificantBits)
  return buf.array()
}

File: gateway/swift/Sources/BridgethingGateway/UUIDBytes.swift
import Foundation

/// UUIDs travel on the wire as 16-byte msgpack `bin`. The schema-generated
/// types expose those fields as `Data`; these initializers translate to and
/// from the native `UUID` type at field boundaries.
public extension UUID {
  init?(data: Data) {
    guard data.count == 16 else { return nil }
    let bytes = [UInt8](data)
    let tuple: uuid_t = (
      bytes[0], bytes[1], bytes[2], bytes[3],
      bytes[4], bytes[5], bytes[6], bytes[7],
      bytes[8], bytes[9], bytes[10], bytes[11],
      bytes[12], bytes[13], bytes[14], bytes[15]
    )
    self.init(uuid: tuple)
  }

  var data: Data {
    withUnsafeBytes(of: uuid) { Data($0) }
  }
}
